[WIP] Add support for gradle resolver

smocherla-brex opened the pull request:
This attempts to add some basic support for a gradle resolver to support usecases like https://github.com/bazel-contrib/rules_jvm_external/issues/864

Right now, this is still a WIP and doesn't address all usecases, but the basic resolution code works for an example

```
smocherla@NLC2L54QQY gradle_resolver % RJE_VERBOSE=true REPIN=1 bazel run @alternative_resolver//:pin                                                        
DEBUG: /private/var/tmp/_bazel_smocherla/d367206b0ffa1c09ec34c0205d1d3720/external/rules_jvm_external+/private/extensions/maven.bzl:351:18: The maven repository 'maven' has contributions from multiple bzlmod modules, and will be resolved together: ["protobuf", "rules_jvm_external_examples_bzlmod"]
DEBUG: /private/var/tmp/_bazel_smocherla/d367206b0ffa1c09ec34c0205d1d3720/external/rules_jvm_external+/private/rules/coursier.bzl:622:18: No file downloaded for org_jetbrains_kotlinx_kotlinx_coroutines_bom_1_9_0, skipping http_file definition
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed target @@rules_jvm_external++maven+alternative_resolver//:pin (194 packages loaded, 3865 targets configured).
INFO: Found 1 target...
Target @@rules_jvm_external++maven+alternative_resolver//:pin up-to-date:
  bazel-out/rules_jvm_external++maven+alternative_resolver/darwin_arm64-fastbuild/bin/pin
  bazel-out/rules_jvm_external++maven+alternative_resolver/darwin_arm64-fastbuild/bin/pin-config.json
INFO: Elapsed time: 0.652s, Critical Path: 0.20s
INFO: 3 processes: 187 action cache hit, 1 internal, 1 darwin-sandbox, 1 worker.
INFO: Build completed successfully, 3 total actions
INFO: Running command line: bazel-out/rules_jvm_external++maven+alternative_resolver/darwin_arm64-fastbuild/bin/pin
Currently: reading parameter file
Resolving dependencies with gradle project: /var/folders/hl/w7gm71y14p34449xg6sm0f980000gn/T/rules_jvm_external12654891016499566404
Downloading https://services.gradle.org/distributions/gradle-8.13-bin.zip
..................................................................................................................................

> Task :resolveDependencies
Dependency graph dumped to: /var/folders/hl/w7gm71y14p34449xg6sm0f980000gn/T/rules_jvm_external12654891016499566404/resolved-dependencies.json

[Incubating] Problems report is available at: file:///private/var/folders/hl/w7gm71y14p34449xg6sm0f980000gn/T/rules_jvm_external12654891016499566404/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.13/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 19s
1 actionable task: 1 executed
Currently: downloading dependencies
Downloading: https://mvnrepository.com/artifact/org/jetbrains/kotlin/kotlin-stdlib/2.0.0/kotlin-stdlib-2.0.0.jar
Downloading: https://mvnrepository.com/artifact/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.9.0/kotlinx-coroutines-core-1.9.0.jar
Downloading: https://mvnrepository.com/artifact/org/jetbrains/annotations/23.0.0/annotations-23.0.0.jar
Downloading: https://mvnrepository.com/artifact/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.jar
Downloading: https://mvnrepository.com/artifact/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.9.0/kotlinx-coroutines-core-jvm-1.9.0.jar
[WARNING]: downloader: Authentication failed for mvnrepository.com. Results from this host will be missing from the generated lock file.
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.9.0/kotlinx-coroutines-core-jvm-1.9.0.jar
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.9.0/kotlinx-coroutines-core-1.9.0.jar
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/kotlin/kotlin-stdlib/2.0.0/kotlin-stdlib-2.0.0.jar
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/annotations/23.0.0/annotations-23.0.0.jar
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.jar
Downloading: https://repo.kotlin.link/org/jetbrains/kotlin/kotlin-stdlib/2.0.0/kotlin-stdlib-2.0.0.jar
Downloading: https://repo.kotlin.link/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.9.0/kotlinx-coroutines-core-1.9.0.jar
Downloading: https://repo.kotlin.link/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.9.0/kotlinx-coroutines-core-jvm-1.9.0.jar
Downloading: https://repo.kotlin.link/org/jetbrains/annotations/23.0.0/annotations-23.0.0.jar
Downloading: https://repo.kotlin.link/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.jar
Downloading: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.jar
Downloading: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.9.0/kotlinx-coroutines-core-jvm-1.9.0.jar
Downloading: https://repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib/2.0.0/kotlin-stdlib-2.0.0.jar
Downloading: https://repo1.maven.org/maven2/org/jetbrains/annotations/23.0.0/annotations-23.0.0.jar
Downloading: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.9.0/kotlinx-coroutines-core-1.9.0.jar
Downloading: https://mvnrepository.com/artifact/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.pom
Downloading: https://maven.pkg.jetbrains.space/mipt-npm/p/sci/maven/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.pom
Downloading: https://repo.kotlin.link/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.pom
Downloading: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-bom/1.9.0/kotlinx-coroutines-bom-1.9.0.pom
Currently: building lock file

```

This uses gradle tooling API https://docs.gradle.org/current/userguide/tooling_api.html which is the recommended way to integrate with gradle, and it requires setting up a fake gradle project with a task that just resolves the dependencies to a JSON file which we try to consume and use in the resolver.

Need to handle exclusions, unsafe cache, conflict resolution and verify the multiplatform usecases work. Flesh out more unit tests and make sure the example also fully works.

shs96c commented:
This is great to see! Thank you!

@smocherla-brex, I've had a stalled branch to do this exact same thing for months. Would you like to work together on this? I've put up my WIP as #1360, but I'm more than happy to drop it in favour of your work. Please feel free to lift anything you think is useful from my WIP PR!

smocherla-brex commented:
@shs96c Thanks for sharing your in-progress changes. I'm happy to help get this over the line. I'll take a look at your PR and see if I can leverage some of it in here later this week when I have cycles and hopefully get it in a place where it can be open for review.

smocherla-brex commented:
Update:
- I have a lot of the tests passing now. Some are still failing and need to fix - related to handling BOMs and Kotlin Multiplatform artifacts.
- The tests are slow because Gradle incurs 10-15 seconds in setup for every test case, so I had to bump the timeout.
- Need to cleanup the code and add comments especially in the tooling API.

I'm hoping to get the above unblocked and the tests passing in the next few days and will open it up for review.

shs96c commented:
@smocherla-brex, I really appreciate all the work you're doing on this. Thank you! When you're ready for review, just let us know. If you need a hand with anything, just ask :)

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleProject.java:
```diff
@@ -0,0 +1,108 @@
+        .addProgressListener(new GradleProgressListener(eventListener))
+        .setStandardError(System.err)
+        .withArguments(arguments)
+      //  .setJvmArguments("-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5005") // uncomment if you want to debug the plugin itself
```
This can be uncommented and be used with `bazel run` to then debug the plugin itself (as it runs with the gradle daemon and not the bazel process). Rather needed to debug some complex issues.

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/models/Exclusion.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Exclusion models a local exclusion to be applied to a dependency in build.gradle.kts
+ */
+public interface Exclusion {
```
This is one of many interfaces and implementations needed to work with the gradle ToolingModelBuilder api to properly serialize the dependency information that's generated in the plugin.

smocherla-brex commented on MODULE.bazel:
```diff
@@ -646,6 +665,24 @@ dev_maven.install(
+    name = "regression_testing_gradle",
+    artifacts = [
+        # https://github.com/bazel-contrib/rules_jvm_external/issues/909
+        "androidx.compose.foundation:foundation-layout:1.5.0-beta01",
```
Added a couple of regression tests for the issues linked above/below.

smocherla-brex commented on tests/com/github/bazelbuild/rules_jvm_external/resolver/gradle/BUILD.bazel:
```diff
@@ -0,0 +1,58 @@
+
+java_test(
+    name = "GradleResolverTest",
+    timeout = "eternal",
```
Increased the timeout here since the target takes about 5-6 minutes to complete as using an isolated instance of gradle per test and running things serially adds up quite a bit. We could look into test sharding if this is unacceptably slow.

smocherla-brex commented:
This should be ready for review now @shs96c. It's possible gradle implementation may not be optimal (as I'm not a gradle expert) but all tests are passing and I also added some regression tests and added comments where relevant for context. There's one job failure in `rules-jvm-external` but that seems potentially unrelated (`kt_jvm_export`) to this PR.

shs96c reviewed (changes requested):
This is great stuff. Thank you so much for working on it. The only real blocker is the comment about the `exclusions` being included in the handlebars context as a `Stream` rather than a collection.

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -0,0 +1,458 @@
+    ResolvedComponentResult root = result.getRoot();
+
+    if (isVerbose()) {
+      System.err.println("Dependency graph: ");
```
Not very obvious, but this goes in tandem with https://github.com/bazel-contrib/rules_jvm_external/pull/1357/files#diff-dde946d96e6addc3d912a0369866caa003e8dd2260d2043b9021d62983347eaeR197 to print the dependency graph. 

Output looks as follows:
```
.Dependency graph: 
  → com.example:base:4.0.0
    → com.example:odd-deps:18.0
  → com.example:diff:2.0.0
    → com.example:odd-deps:18.0
```

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleBuildScriptGenerator.java:
```diff
@@ -0,0 +1,235 @@
+    handlebars.registerHelper(
+        "lowerCase",
+        (context, options) -> {
+          if (context == null) return "";
```
Actually, I was using this function earlier in the handlebars template but removed it at some point, so no longer needed and I removed this function.

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleBuildScriptGenerator.java:
```diff
@@ -0,0 +1,235 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
```
Updated.

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleProject.java:
```diff
@@ -0,0 +1,108 @@
+
+    private final Path projectDir;
+    private final Path gradleCacheDir;
+    private ProjectConnection connection;
```
updated

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleProject.java:
```diff
@@ -0,0 +1,108 @@
+        List<String> arguments = gradleProperties.entrySet().stream()
+                .map(entry -> "-P" + entry.getKey() + "=" + entry.getValue())
+                .collect(Collectors.toList());
+        if(this.initScript != null) {
```
updated

smocherla-brex commented on MODULE.bazel:
```diff
@@ -826,8 +863,10 @@ use_repo(
+    "regression_testing_gradle",
     "regression_testing_maven",
     "unpinned_regression_testing_coursier",
+    "unpinned_regression_testing_gradle",
```
Removed

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -0,0 +1,458 @@
+      // this is a new artifact yet unvisited
+      info = new GradleResolvedDependencyImpl();
+    }
+    info.setGroup(component.getModuleVersion().getGroup());
```
Good catcg, moved it inside the block where the object is created. It's redundant.

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -0,0 +1,458 @@
+    List<GradleResolvedDependency> children = new ArrayList<>();
+
+    for (DependencyResult dep : component.getDependencies()) {
+      if (!(dep instanceof ResolvedDependencyResult)) continue;
```
updated

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -0,0 +1,458 @@
+  }
+
+  private String extractExtension(File file) {
+    String name = file.getName();
```
updated

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -0,0 +1,458 @@
+  }
+
+  private boolean isVerbose() {
+    return System.getenv("RJE_VERBOSE") != null && System.getenv("RJE_VERBOSE").equals("true");
```
updated

smocherla-brex commented:
> This is great stuff. Thank you so much for working on it. The only real blocker is the comment about the exclusions being included in the handlebars context as a Stream rather than a collection.

Thanks, updated the PR and addressed the feedback you shared. Also added a small update to the README about the new resolver.

smocherla-brex commented:
I've been looking at improving performance of the resolver as it's extremely slow right now (on our monorepo, it takes 15 mins+ which isn't great). My hunch for one of the reasons (not everything) based on testing so far is that it's really slow right now due to the serial nature of artifact resolution in the plugin for various artifact views (javadocs, source jars) and additionally serially downloading POMs for every artifact rather than batching. I'm going to try make some of those resolution async and see if it helps the performance.

shs96c commented:
@smocherla-brex, I plan on merging this on Monday since it LGTM, but I'd like the other committers to also have a final chance to review. Thank you for working on this!
